src/framework/rng.hpp

```diff
@@ -18,117 +18,83 @@
-   */
-  int_t rand_int(int_t a, int_t b);
-  uint_t rand_int(uint_t a, uint_t b);
-
-  /**
-   * Generate a pseudo random integer from a a discrete distribution
-   * constructed from an input vector of probabilities for [0,..,n-1]
-   * where n is the lenght of the vector. If this vector is not normalized
-   * it will be scaled when it is converted to a discrete_distribution
-   * @param probs the vector of probabilities
-   * @return the generated integer
-   */
-  uint_t rand_int(const std::vector<double> &probs);
-
-  /**
-   * Default constructor initialize RNG engine with a random seed
-   */
-  RngEngine() {
-    std::random_device rd;
-    rng.seed(rd());
-  };
 
-  /**
-   * Seeded constructor initialize RNG engine with a fixed seed
-   * @param seed integer to use as seed for mt19937 engine
-   */
-  explicit RngEngine(uint_t seed) { rng.seed(seed); };
+  //-----------------------------------------------------------------------
+  // Constructors
+  //-----------------------------------------------------------------------
+
+  // Default constructor initialize RNG engine with a random seed
+  RngEngine() { set_random_seed(); }
+
+  // Seeded constructor initialize RNG engine with a fixed seed
+  explicit RngEngine(size_t seed) { set_seed(seed); }
+
+  //-----------------------------------------------------------------------
+  // Set fixed or random seed for RNG
+  //-----------------------------------------------------------------------
 
+  // Set random seed for the RNG engine
+  void set_random_seed() { 
+    std::random_device rd;
+    set_seed(rd());
+  }
 
   // Set a fixed seed for the RNG engine
-  void set_seed(uint_t seed) { rng.seed(seed); };
+  void set_seed(size_t seed) { rng.seed(seed); }
+  
+  //-----------------------------------------------------------------------
+  // Sampling methods
+  //-----------------------------------------------------------------------
+
+  // Generate a uniformly distributed pseudo random real in the half-open
+  // interval [a,b)
+  double rand(double a, double b) {
+    return std::uniform_real_distribution<double>(a, b)(rng);
+  }
+
+  // Generate a uniformly distributed pseudo random real in the half-open
+  // interval [0,b)
+  double rand(double b) {
+    return rand(double(0), b);
+  };
+
+  // Generate a uniformly distributed pseudo random real in the half-open
+  // interval [0,1)
+  double rand() {return rand(0, 1); };
+
+  // Generate a uniformly distributed pseudo random integer in the closed
+  // interval [a,b]
+  template <typename IntType>
+  IntType rand_int(IntType a, IntType b) {
+    return std::uniform_int_distribution<IntType>(a, b)(rng);
+  }
+
+  // Generate a pseudo random integer from a a discrete distribution
+  // constructed from an input vector of probabilities for [0,..,n-1]
+  // where n is the lenght of the vector. If this vector is not normalized
+  // it will be scaled when it is converted to a discrete_distribution
+  template <typename FloatType>
```

Comment:
What about checking `FloatType` is actually a floating point type?
```c++
template<typename FloatType,
         typename = typename std::enable_if<std::is_floating_point<FloatType>::value >::type>
size_t rand_int(const std::vector<FloatType> &probs) {
        return std::discrete_distribution<size_t>(probs.begin(), probs.end())(rng);
}

```
(I really hope that Concepts and `if constexpr` will help us with these things)